### PR TITLE
Fix schema inference in case of empty messages in Protobuf/CapnProto formats

### LINKED ISSUE #39051: Protobuf schema inference can make an illegal table
**Describe what's wrong**

Protobuf schema inference will yield an empty `Tuple()` when faced with an empty nested Message. ClickHouse will allow the creation of an empty `Tuple()` column despite this being illegal in a regular `CREATE TABLE` statement (see #39049). Inserts into the resulting table yield `Code: 1001. DB::Exception: std::out_of_range: vector. (STD_EXCEPTION)`

**Does it reproduce on recent release?**

Reproduces on `clickhouse-common-static-22.6.3.35` on Linux.

**How to reproduce**

```
% cat Empty.proto 
syntax = "proto3";

package demo;

message Empty {}

message Foo {
  uint32 i = 1;
  Empty  e = 2;
}

% clickhouse-common-static-22.6.3.35/usr/bin/clickhouse local
ClickHouse local version 22.6.3.35 (official build).

server :) CREATE TABLE foo  ENGINE=File('Protobuf', 'nonexist') SETTINGS format_schema='Empty.proto:Foo';

CREATE TABLE foo
ENGINE = File('Protobuf', 'nonexist')
SETTINGS format_schema = 'Empty.proto:Foo'

Query id: 88854aa7-4129-4675-bb0a-da17fcba1148

Ok.

0 rows in set. Elapsed: 0.001 sec. 

server :) desc table foo;

DESCRIBE TABLE foo

Query id: 0e438a5b-cb53-42d5-9ae9-47050c3f5423

┌─name─┬─type────┬─default_type─┬─default_expression─┬─comment─┬─codec_expression─┬─ttl_expression─┐
│ i    │ UInt32  │              │                    │         │                  │                │
│ e    │ Tuple() │              │                    │         │                  │                │
└──────┴─────────┴──────────────┴────────────────────┴─────────┴──────────────────┴────────────────┘

2 rows in set. Elapsed: 0.001 sec. 

server :) CREATE TABLE foo_mt AS foo Engine=MergeTree() ORDER BY i;

CREATE TABLE foo_mt AS foo
ENGINE = MergeTree
ORDER BY i

Query id: e07c8e3d-47c4-4a69-bdf3-b3ad4b016fc5

Ok.

0 rows in set. Elapsed: 0.009 sec. 

server :) desc foo_mt

DESCRIBE TABLE foo_mt

Query id: 5371e326-646a-4a13-a8fe-b406d55574da

┌─name─┬─type────┬─default_type─┬─default_expression─┬─comment─┬─codec_expression─┬─ttl_expression─┐
│ i    │ UInt32  │              │                    │         │                  │                │
│ e    │ Tuple() │              │                    │         │                  │                │
└──────┴─────────┴──────────────┴────────────────────┴─────────┴──────────────────┴────────────────┘

2 rows in set. Elapsed: 0.000 sec. 

server :) insert into foo_mt (i, e) values (0, tuple());

INSERT INTO foo_mt (i, e) FORMAT Values

Query id: 4e7b0226-745a-40ca-8a65-77354a64947b


0 rows in set. Elapsed: 0.159 sec. 

Received exception:
Code: 1001. DB::Exception: std::out_of_range: vector. (STD_EXCEPTION)

server :) insert into foo_mt (i) values (0);

INSERT INTO foo_mt (i) FORMAT Values

Query id: 8240afed-3e05-4789-9587-4db3a6f9f4b6


0 rows in set. Elapsed: 0.000 sec. 

Received exception:
Code: 1001. DB::Exception: std::out_of_range: vector. (STD_EXCEPTION)
```

**Expected behavior**

Either the table should not be created (protobuf schema inference should fail loudly) or insert to the table should succeed.

** Code context **

The empty `Tuple()` inference happens here: https://github.com/ClickHouse/ClickHouse/blob/v22.6.3.35-stable/src/Formats/ProtobufSerializer.cpp#L3489-L3508

### PR DESCRIPTION
**Pull request by Avogar**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix bug in schema inference in case of empty messages in Protobuf/CapnProto formats that allowed to create column with empty `Tuple` type. Closes https://github.com/ClickHouse/ClickHouse/issues/39051
Add 2 new settings `input_format_{protobuf/capnproto}_skip_fields_with_unsupported_types_in_schema_inference` that allow to skip fields with unsupported types while schema inference for Protobuf and CapnProto formats

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by ekpdt** [comment]
Our current workaround for this bug is to use schema inference to get a preliminary schema on our (extremely) complex Protobuf messages and then edit the inferred schema to remove illegal empty tuples.

This patch would make it impossible for us to use schema inference at all, which is unfortunate.

Would you consider allowing some way to DESCRIBE a protobuf-based schema via the File engine but prohibit creating any actual tables? I admit this seems like a weird solution because CH will still create a schema that is illegal.

If you integrate this patch, we will probably pre-process our .proto files to remove empty messages.

**Comment by Avogar** [comment]
> Our current workaround for this bug is to use schema inference to get a preliminary schema on our (extremely) complex Protobuf messages and then edit the inferred schema to remove illegal empty tuples.
> 
> This patch would make it impossible for us to use schema inference at all, which is unfortunate.
> 
> Would you consider allowing some way to DESCRIBE a protobuf-based schema via the File engine but prohibit creating any actual tables? I admit this seems like a weird solution because CH will still create a schema that is illegal.
> 
> If you integrate this patch, we will probably pre-process our .proto files to remove empty messages.

What if I add special setting like `input_format_protobuf_skip_columns_with_unsupported_types_in_schema_inference` to silently skip bad messages (like empty tuples) instead of throwing exceptions? Will it help?

**Comment by ekpdt** [comment]
>What if I add special setting like input_format_protobuf_skip_columns_with_unsupported_types_in_schema_inference to silently skip bad messages (like empty tuples) instead of throwing an exception? Will it help?

That would be perfect, thanks.

File: src/Core/Settings.h [hunk 1]
```diff
@@ -697,6 +697,8 @@ static constexpr UInt64 operator""_GiB(unsigned long long value)
     M(Bool, input_format_csv_use_best_effort_in_schema_inference, true, "Use some tweaks and heuristics to infer schema in CSV format", 0) \
     M(Bool, input_format_tsv_use_best_effort_in_schema_inference, true, "Use some tweaks and heuristics to infer schema in TSV format", 0) \
     M(Bool, input_format_parquet_skip_columns_with_unsupported_types_in_schema_inference, false, "Allow to skip columns with unsupported types while schema inference for format Parquet", 0) \
+    M(Bool, input_format_protobuf_skip_fields_with_unsupported_types_in_schema_inference, false, "Allow to skip fields with unsupported types while schema inference for format Protobuf", 0) \
```
Comment (rschu1ze) [inline src/Core/Settings.h:700]: (Minor: Dropping the "Allow to " prefix in l. 699-703 would loose no information but make the description a bit more concise)

**Review by rschu1ze — APPROVED** [review]

**Comment by tavplubix** [comment]
New test is flaky with Ordinary database: https://s3.amazonaws.com/clickhouse-test-reports/0/8e98b7d47dbbcfe88c8d420b35794854e93a6361/stateless_tests__release__databaseordinary_.html

https://github.com/ClickHouse/ClickHouse/pull/39623

**Pull request merged**